### PR TITLE
Reorganise configuration :running_shirt_with_sash: 

### PR DESCRIPTION
**Pull request by adrien-zinger**
Yes, this is what you think.

This PR is the proposal for a reorganization/documentation of the settings/configuration. I think they are still some work to do, but I would like to give this job to another person :rofl:

The decision is: _make the configuration a dependency to inject._ In order to be very flexible in test, we should use the configuration as a big dependency. That means that we clone the configuration for each worker (just what the worker need :slightly_smiling_face: not more!)

## Default config
I've created two files:
* default.rs
* default_testing.rs

These files are not destined to be modified every day, I also think that the `default_testing` should never be modified. I just define the default value of the hard-coded configuration for both test and normal configurations structures.

In the test, I used the power of the [default design pattern](https://rust-unofficial.github.io/patterns/idioms/default.html), much cleaner. So now if you want to create a test, you can use the default configuration as defined in `default_testing.rs` and change only what you want to change.

## From traits
I wanted to avoid creating a configuration manually because of the verbosity and also because we don't care for the most of the time of what do we need to create a configuration. If a crate use another crate, we can just use the `from pattern` and avoid a lot of code duplication.

### CODE REVIEW

**Review by yvan-sraka — DISMISSED** [review]
That's really a bunch of works and a great improvement of codebase quality and general organization :tada:!

**Comment by AureliaDolo** [comment]
It's based on this #2211 ?

**Review by AureliaDolo — DISMISSED** [review]

File: massa-models/src/node_configuration/compact_config.rs [hunk 1]
```diff
@@ -42,11 +25,33 @@ pub struct CompactConfig {
...
+            thread_count: THREAD_COUNT,
+            t0: *T0,
+            delta_f0: DELTA_F0,
+            operation_validity_periods: OPERATION_VALIDITY_PERIODS,
+            periods_per_cycle: PERIODS_PER_CYCLE,
+            pos_lookback_cycles: POS_LOOKBACK_CYCLES,
+            pos_lock_cycles: POS_LOCK_CYCLES,
+            block_reward: *BLOCK_REWARD,
+            roll_price: *ROLL_PRICE,
+            max_block_size: EVENT_ID_SIZE_BYTES as u32,
```
Comment (adrien-zinger) [inline massa-models/src/node_configuration/compact_config.rs:50]: I miss that! Need to change for MAX_BLOCK_SIZE

File: massa-consensus-exports/src/settings.rs [hunk 2]
```diff
@@ -1,20 +1,59 @@
...
+/// # Channels
+///
+/// The following file contains the definition of the Channels structures used in
+/// the current module.
+///
+/// # Testing feature
+///
+/// In unit test your allowed to use the `testing` feature flag that will
+/// use the default values from `/node_configuration/default_testing.rs` in the
+/// `massa-models` crate sources.
```
Comment (AureliaDolo) [inline massa-consensus-exports/src/settings.rs:46]: That would be great duplicated is some file in the wiki, or in a readme ?

**Review by AureliaDolo — COMMENTED** [review]
It's quite not clear for what's the difference between xxxSettings and xxxConfig yet

**Review by yvan-sraka — DISMISSED** [review]

File: massa-consensus-worker/src/tests/test_block_graph.rs [hunk 3]
```diff
@@ -363,12 +364,30 @@ pub async fn test_get_ledger_at_parents() {
...
             (hash_genesist1, export_genesist1),
-            (get_dummy_block_id("blockp1t0"), (&blockp1t0).into()),
-            (get_dummy_block_id("blockp1t1"), (&blockp1t1).into()),
-            (get_dummy_block_id("blockp2t0"), (&blockp2t0).into()),
-            (get_dummy_block_id("blockp2t1"), (&blockp2t1).into()),
-            (get_dummy_block_id("blockp3t0"), (&blockp3t0).into()),
-            (get_dummy_block_id("blockp3t1"), (&blockp3t1).into()),
+            (
+                get_dummy_block_id("blockp1t0"),
+                ExportActiveBlock::from(&blockp1t0),
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-worker/src/tests/test_block_graph.rs:369]: It looks like clippy doesn't understand the borrowing with the into in that case, I don't know why! :thinking:

File: massa-consensus-exports/src/settings.rs [hunk 4]
```diff
@@ -418,64 +327,323 @@ pub struct ConsensusChannels {
...
+        massa_models::init_serialization_context(massa_models::SerializationContext::default());
+        let tempdir = tempfile::tempdir().expect("cannot create temp dir for the ledger path");
+        let path_buf = tempdir.path().to_path_buf();
+        Self {
+            temp_files: TempFiles {
+                temp_dir: vec![tempdir],
+                temp_files: vec![],
+            },
+            // reset genesis timestamp because we are in test mode that can take a while to process
+            genesis_timestamp: MassaTime::now().unwrap(),
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs:540]: You don't like unwraps? :rofl:

File: massa-consensus-exports/src/settings.rs [hunk 5]
```diff
@@ -301,32 +173,69 @@ pub struct ConsensusConfig {
     pub max_item_return_count: usize,
     /// If we want to generate blocks.
     /// Parameter that shouldn't be defined in prod.
-    #[serde(skip, default = "Default::default")]
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs:304]: Is that used?

File: massa-consensus-exports/src/settings.rs (hunk 2 above)
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs]: Ok, I think it's already in the rust doc, It would be great to copy it in user doc also :+1:

**Review by yvan-sraka — DISMISSED** [review]

File: massa-bootstrap/src/tests/tools.rs [hunk 6]
```diff
@@ -69,7 +70,7 @@ pub fn get_bootstrap_config(bootstrap_public_key: PublicKey) -> BootstrapSetting
         max_bootstrap_message_size: 100000000,
         max_bootstrap_pos_entries: 1000,
         max_bootstrap_pos_cycles: 5,
-        max_block_endorsements: 8,
+        endorsement_count: 8,
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-bootstrap/src/tests/tools.rs:73]: It's the same, it's just the name you choose.

File: massa-consensus-exports/src/settings.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-consensus-exports/src/settings.rs:304]: Yep, this field does not appear in the config file, hence it is inferred here.

File: massa-consensus-worker/src/tests/test_block_graph.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-consensus-worker/src/tests/test_block_graph.rs:369]: arg

**Review by damip — COMMENTED** [review]
Some comments:
* a lot of the lazy_static are actually not necessary. Sorry I prefer to avoid global variables when not absolutely necessary.
* trying to have the same config structure (ex: pool settings) loaded from file AND passed to the module is often inadequate. I vote for having structures in massa-node representing only the config layout within the config file, and then having a unique PoolConfig containing all parameters (both thread_count and max_pool_size_per_thread) at the same level, so no need for an extra layer of lazy static stuff. PoolConfig would simply be filled by the node before launching the Pool controller, using values coming from the settings file and/or constants

**Comment by yvan-sraka** [comment]
> That's really a bunch of works and a great improvement of codebase quality and general organization tada!

And I really appreciate and thank you for the documentation effort you put in that PR! Everything is so much easier to find in code, read and understand :)

**Review by damip — APPROVED** [review]
It's a welcome improvement on what we have and we can already merge it. Please open a followup for removing useless statics and make config passing uniform (see my previous comment)

File: massa-consensus-worker/src/tests/tools.rs [hunk 7]
```diff
@@ -572,78 +612,22 @@ pub fn get_creator_for_draw(draw: &Address, nodes: &Vec<PrivateKey>) -> PrivateK
...
+//    let thread_count: u8 = 2;
+//    let max_block_size: u32 = 3 * 1024 * 1024;
+//    let max_operations_per_block: u32 = 1024;
+//    let tempdir = tempfile::tempdir().expect("cannot create temp dir");
+//
+//    // Init the serialization context with a default,
+//    // can be overwritten with a more specific one in the test.
+//    massa_models::init_serialization_context(massa_models::SerializationContext::default());
+//
+//}
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-worker/src/tests/tools.rs:630]: done

File: massa-consensus-exports/src/settings.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs:304]: reset

File: massa-models/src/node_configuration/mod.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * Nevertheless the disign is more like using `massa_config::CONST_VALUE`
+ * and defining in `Cargo.toml` if we are testing or not
+ *
+```ignore
+[dependencies]
+    massa_config = { path = "../massa-config" }
+[dev-dependencies]
+    massa_config = { path = "../massa-config", features = ["testing"] }
+```
+ */
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-models/src/node_configuration/mod.rs]: I guess that is done :slightly_smiling_face:

File: massa-models/src/context.rs [hunk 9]
```diff
@@ -55,17 +55,16 @@ where
...
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct SerializationContext {
-    pub max_block_operations: u32,
-    pub max_block_endorsements: u32,
-    pub parent_count: u8,
+    pub max_operations_per_block: u32,
+    pub endorsement_count: u32,
+    pub thread_count: u8,
     pub max_block_size: u32,
     // TODO: all field bellow should be removed since there are constant defined in bootstrap::settings
```
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-models/src/context.rs]: done!

File: massa-consensus-exports/src/settings.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrien-zinger) [reply massa-consensus-exports/src/settings.rs:540]: I make a commit just for that :wink:

File: massa-consensus-worker/src/tests/scenario_block_creation.rs [hunk 10]
```diff
@@ -1,60 +1,44 @@
...
     ledger.insert(
         address_2,
         LedgerData::new(Amount::from_str("1000").unwrap()),
     );
-    let ledger_file = generate_ledger_file(&ledger);
-    let staking_keys: Vec<PrivateKey> = vec![priv_1, priv_2];
-
-    // init roll cont
+    let mut cfg = ConsensusConfig {
+        genesis_timestamp: MassaTime::from_str("1633301290000").unwrap(),
```
[the remark this answers is not in the document]
Comment (AureliaDolo) [reply massa-consensus-worker/src/tests/scenario_block_creation.rs:38]: see  #2301

**Review by yvan-sraka — APPROVED** [review]

**Review by damip — APPROVED** [review]

**Comment by damip** [comment]
let's merge this

**Pull request merged**